lib/collection/src/collection.rs

```diff
@@ -1039,22 +1039,120 @@ impl Collection {
         Ok(points)
     }
 
-    #[allow(unreachable_code, clippy::diverging_sub_expression)]
     pub async fn update_params_from_diff(
         &self,
         params_diff: CollectionParamsDiff,
     ) -> CollectionResult<()> {
         let mut config = self.config.write().await;
         config.params = params_diff.update(&config.params)?;
-        self.handle_replica_changes(todo!("supply replica changes"));
+        // TODO supply replica changes
+        let changes = HashSet::new();
+        self.handle_replica_changes(changes).await?;
         Ok(())
     }
 
-    pub fn handle_replica_changes(&self, _replica_changes: HashSet<replica_set::Change>) {
-        // TODO: remove or add replicas. In case of replica addition:
-        // 1. Create and mark them as inactive
-        // 2. Copy data
-        // 3. Mark them as active
+    /// Handle replica changes
+    ///
+    /// add and remove replicas from replica set
+    pub async fn handle_replica_changes(
+        &self,
+        replica_changes: HashSet<Change>,
+    ) -> CollectionResult<()> {
+        if replica_changes.is_empty() {
+            return Ok(());
+        }
+
+        let mut shard_holder = self.shards_holder.write().await;
+
+        // all existing replica sets in the collection
+        let mut replica_sets = Vec::new();
+        for shard in shard_holder.all_shards() {
+            match shard {
+                Shard::Local(_) => {
+                    log::error!("A replicated collection should not contain a local shard")
+                }
+                Shard::Proxy(_) => {
+                    log::error!("A replicated collection should not contain a proxy shard")
+                }
+                Shard::ForwardProxy(_) => {
+                    log::error!("A replicated collection should not contain a forward proxy shard")
+                }
+                Shard::Remote(_) => {
+                    log::error!("A replicated collection should not contain a remote shard")
+                }
+                Shard::ReplicaSet(replica_set) => replica_sets.push(replica_set.shard_id),
+            }
```

Comment:
```suggestion
            match shard {
                Shard::ReplicaSet(replica_set) => replica_sets.push(replica_set.shard_id),
                shard => log::error!("A replicated collection should not contain a {}", shard.variant_name())
            }
```
And implement method `variant_name` on enum `Shard`

I think this will be quite useful in many similar error handling cases

Reply:
I was unsure about such approach as we have been performing exhaustive pattern match manually so far on the `Shard` enum.
I thought we wanted to get the help from the compiler when adding a new variant.
WDYT?
Not against your proposal :)

Reply:
Yeah the point about getting alerted in case of new variants is a good one. In that case we can do something like:
```rust
shard @ Shard::Local(_) | Shard::Proxy(_) | Shard::ForwardProxy(_) | Shard::Remote(_) =>  log::error!("A replicated collection should not contain a {}", shard.variant_name())
```

Reply:
Thanks for the suggestion :ok_hand: 

https://github.com/qdrant/qdrant/pull/1058/commits/26fec329f9a3678b8085046ffc98fd51e993b917